projects/components/src/table/data/table-cdk-data-source.ts

```diff
@@ -274,19 +283,57 @@ export class TableCdkDataSource implements DataSource<TableRow> {
       return of([]);
     }
 
-    return this.tableDataSourceProvider.data
-      .getData(this.buildRequest(columnConfigs, pageEvent, filters, queryProperties))
-      .pipe(
-        tap(response => this.updatePaginationTotalCount(response.totalCount)),
-        map(response => response.data),
-        map(rows => this.paginateRows(rows, pageEvent)),
-        map(TableCdkRowUtil.buildInitialRowStates),
-        map(rows =>
-          this.rowStateChangeProvider.initialExpandAll && TableCdkRowUtil.isFullyExpandable(rows)
-            ? TableCdkRowUtil.expandAllRows(rows)
-            : rows
-        )
-      );
+    return this.hasCachedRowsForPageEvent(pageEvent)
+      ? this.fetchCachedData(pageEvent)
+      : this.fetchNewData(this.buildRequest(columnConfigs, pageEvent, filters, queryProperties));
+  }
+
+  private hasCachedRowsForPageEvent(pageEvent: PageEvent): boolean {
+    if (this.cachedRows.length < pageEvent.pageSize) {
+      // Sanity check if we have enough cached data for what we request
+      return false;
+    }
+
+    const offsetWithinCachedRows = this.calcOffsetWithinCachedRows(pageEvent);
+
+    // Check if requested startOffset + limit is within the cached data
+    return offsetWithinCachedRows >= 0 && offsetWithinCachedRows + pageEvent.pageSize <= this.cachedRows.length;
+  }
+
+  private calcOffsetWithinCachedRows(pageEvent: PageEvent): number {
+    const cachedOffset = this.cachedPosition?.startIndex ?? 0;
+    const currentOffset = pageEvent.pageIndex * pageEvent.pageSize;
+
+    return currentOffset - cachedOffset;
+  }
+
+  private fetchNewData(request: TableDataRequest): Observable<StatefulTableRow[]> {
+    if (this.tableDataSourceProvider.data === undefined) {
+      return of([]);
```

Comment:
this will lead us to absorb errors, right? shouldn't this be an error if the datasource provider is undefined?

Reply:
I know this isn't new change. Just a question in general :)

Reply:
Yeah, the check has been there a while. Pretty sure we never actually hit this.